File: java/src/main/java/com/example/cote/PGMRS/유전_법칙.java
package com.example.cote.PGMRS;

import java.util.*;
import java.util.function.*;

public class 유전_법칙 {

	private static final int MAX_QUERIES_LENGTH = 5;
	private static final int MAX_GENERATION = 16;
	private static final IntUnaryOperator gen =  n -> (int) Math.pow(4, n - 1);

	private static final Random random = new Random(43);
	private static final int TEST_CASES = 10;

	private static String find(int n, int p) {
		if (n == 1) {
			return "Rr";
		}

		int chunk = (int) Math.pow(4, n - 2);
		int group = (p - 1) / chunk;

		if (group == 0) {
			return "RR";
		}

		if (group == 3) {
			return "rr";
		}

		return find(n - 1, p % chunk);
	}

	private static String[] solve(int[][] queries) {
		return Arrays.stream(queries)
			.map(query -> find(query[0], query[1] ))
			.toArray(String[]::new);
	}
	
	public static void main(String... args) {

		for (int it = 0; it < TEST_CASES; it++) {
			int q = random.nextInt(MAX_QUERIES_LENGTH - 1) + 1;
			int[][] queries = new int[q][2];

			for (int i = 0; i < q; i++) {
				int g = random.nextInt(MAX_GENERATION - 2) + 2;
				queries[i][0] = g;
				queries[i][1] = random.nextInt(gen.applyAsInt(g) - 1) + 1;
			}
			System.out.println(Arrays.toString(solve(queries)));
		}

	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/유전_법칙.kt
package com.example.cote.PGMRS;

import kotlin.math.*
import kotlin.random.Random

val MAX_QUERIES_LENGTH = 5
val MAX_GENEATION = 16

val TEST_CASES = 10
val random = Random(43)

fun gen(n: Int) = (4.0).pow(n - 1).toInt()

fun Random.nextQueryLength(): Int 
	= nextInt(MAX_QUERIES_LENGTH - 1) + 1

fun Random.nextQuery(): Pair<Int, Int> {
	val generation = nextInt(MAX_GENEATION - 2) + 2
	val number = nextInt(gen(generation))
	return Pair(generation, number)
}

fun find(generation: Int,  number: Int): String {
	if (generation == 1) {
		return "Rr"
	}

	val chunk = gen(generation - 1)

	return when (number / chunk) {
		0 -> "RR"
		3 -> "rr"
		else -> find(generation - 1, number % chunk)
	}
}

fun solve(queries: List<Pair<Int, Int>>): List<String> {
	return queries.map{(generation, number) -> 
		find(generation, number)
	}
}

fun main() {

	repeat(TEST_CASES) {
		val queries = List(random.nextQueryLength()) {
			random.nextQuery()
		}
		println(solve(queries))
	}
}